File: articles/alonzo-church-lambda-calculus-and-formal-computation/go/church_numeral.go
package main

import "fmt"

func churchApply(n int, f func(int) int, x int) int {
	for i := 0; i < n; i++ {
		x = f(x)
	}
	return x
}

func main() {
	fmt.Printf("church_3_successor_0=%d\n", churchApply(3, func(x int) int { return x + 1 }, 0))
}

File: articles/alonzo-church-lambda-calculus-and-formal-computation/rust/church_numeral.rs
fn church_apply<F>(n: usize, mut f: F, mut x: i32) -> i32
where
    F: FnMut(i32) -> i32,
{
    for _ in 0..n {
        x = f(x);
    }
    x
}

fn main() {
    println!("church_3_successor_0={}", church_apply(3, |x| x + 1, 0));
}
